data-prepper-plugins/mongodb/src/main/java/org/opensearch/dataprepper/plugins/mongo/stream/StreamWorker.java

```diff
@@ -0,0 +1,164 @@
+    private final AcknowledgementSetManager acknowledgementSetManager;
+    private final  PluginMetrics pluginMetrics;
+    private final int defaultFlushBatchSize;
+
+    private final JsonWriterSettings writerSettings = JsonWriterSettings.builder()
+            .outputMode(JsonMode.RELAXED)
+            .objectIdConverter((value, writer) -> writer.writeString(value.toHexString()))
+            .build();
+
+    public static StreamWorker create(final RecordBufferWriter recordBufferWriter,
+                         final AcknowledgementSetManager acknowledgementSetManager,
+                         final MongoDBSourceConfig sourceConfig,
+                         final DataStreamPartitionCheckpoint partitionCheckpoint,
+                         final PluginMetrics pluginMetrics,
+                         final int defaultFlushBatchSize
+    ) {
+        return new StreamWorker(recordBufferWriter, acknowledgementSetManager,
+                sourceConfig, partitionCheckpoint, pluginMetrics, defaultFlushBatchSize);
+    }
+    public StreamWorker(final RecordBufferWriter recordBufferWriter,
+                        final AcknowledgementSetManager acknowledgementSetManager,
+                        final MongoDBSourceConfig sourceConfig,
+                        final DataStreamPartitionCheckpoint partitionCheckpoint,
+                        final PluginMetrics pluginMetrics,
+                        final int defaultFlushBatchSize
+                        ) {
+        this.recordBufferWriter = recordBufferWriter;
+        this.sourceConfig = sourceConfig;
+        this.partitionCheckpoint = partitionCheckpoint;
+        this.acknowledgementSetManager = acknowledgementSetManager;
+        this.pluginMetrics = pluginMetrics;
+        this.defaultFlushBatchSize = defaultFlushBatchSize;
+        this.successItemsCounter = pluginMetrics.counter(SUCCESS_ITEM_COUNTER_NAME);
+        this.failureItemsCounter = pluginMetrics.counter(FAILURE_ITEM_COUNTER_NAME);
+    }
+
+    private MongoCursor<ChangeStreamDocument<Document>> getChangeStreamCursor(final MongoCollection<Document> collection,
+                            final String resumeToken
+                            ) {
+        if (resumeToken == null) {
+            return collection.watch().fullDocument(FullDocument.UPDATE_LOOKUP).iterator();
+        } else {
+            return collection.watch().fullDocument(FullDocument.UPDATE_LOOKUP).resumeAfter(BsonDocument.parse(resumeToken)).maxAwaitTime(60, TimeUnit.SECONDS).iterator();
+        }
+    }
+
+    private boolean shouldWaitForExport(final StreamPartition streamPartition) {
+        final StreamProgressState progressState = streamPartition.getProgressState().get();
+        // when export is complete the global load status is created
+        final Optional<StreamLoadStatus> loadStatus = partitionCheckpoint.getGlobalStreamLoadStatus();
+        return progressState.shouldWaitForExport() && loadStatus.isEmpty();
```

Comment:
Curious in which case will we have different true/false values between `progressState.shouldWaitForExport()` and `loadStatus.isEmpty()`, e.g. one of them is true and the other is false?

Reply:
The Global stream load status is created when export is complete by export worker. If there is no export then global status is not created and loadstatus would be empty. `progressState.shouldWaitForExport()` is set to true if pipeline is configured for export or export and stream. If both are true, then stream worker will wait. If one of them is false (shouldWaitForExport is false, load status is always empty) stream will continue.